library/pod-security-policy/users/template.yaml

```diff
@@ -132,158 +132,341 @@ spec:
                         description: "The maximum group ID in the range, inclusive."
   targets:
     - target: admission.k8s.gatekeeper.sh
-      rego: |
-        package k8spspallowedusers
+      code:
+      - engine: K8sNativeValidation
+        source:
+          variables:
+          - name: containers
+            expression: 'has(variables.anyObject.spec.containers) ? variables.anyObject.spec.containers : []'
+          - name: initContainers
+            expression: 'has(variables.anyObject.spec.initContainers) ? variables.anyObject.spec.initContainers : []'
+          - name: ephemeralContainers
+            expression: 'has(variables.anyObject.spec.ephemeralContainers) ? variables.anyObject.spec.ephemeralContainers : []'
+          - name: exemptImagePrefixes
+            expression: |
+              !has(variables.params.exemptImages) ? [] :
+                variables.params.exemptImages.filter(image, image.endsWith("*")).map(image, string(image).replace("*", ""))
+          - name: exemptImageExplicit
+            expression: |
+              !has(variables.params.exemptImages) ? [] : 
+                variables.params.exemptImages.filter(image, !image.endsWith("*"))
+          - name: exemptImages
+            expression: |
+              (variables.containers + variables.initContainers + variables.ephemeralContainers).filter(container,
+                container.image in variables.exemptImageExplicit ||
+                variables.exemptImagePrefixes.exists(exemption, string(container.image).startsWith(exemption)))
+          - name: badContainers
+            expression: |
+              (variables.containers + variables.initContainers + variables.ephemeralContainers).filter(container,
+                !(container.image in variables.exemptImages))
+          - name: podRunAsUser
+            expression: |
+              variables.anyObject.kind == "Pod" ? 
+                (has(variables.anyObject.spec.securityContext) && has(variables.anyObject.spec.securityContext.runAsUser) ? variables.anyObject.spec.securityContext.runAsUser : null)
+              : null
+          - name: missingRequiredRunAsUserContainers
+            expression: |
+              variables.badContainers.filter(container, 
+                has(variables.params.runAsUser) && has(variables.params.runAsUser.rule) && (variables.params.runAsUser.rule == "MustRunAs") &&
+                !((has(container.securityContext) && has(container.securityContext.runAsUser)) || 
+                  (variables.podRunAsUser != null)))
+          - name: missingRequiredRunAsNonRootContainers
+            expression: |
+              variables.badContainers.filter(container, 
+                has(variables.params.runAsUser) && has(variables.params.runAsUser.rule) && (variables.params.runAsUser.rule == "MustRunAsNonRoot") && ((has(container.securityContext) && !(has(container.securityContext.runAsUser) || has(container.securityContext.runAsNonRoot))) ?
```

Comment:
We need to check that runAsNonRoot is both defined and equal to `true` (this is implied in the Rego check, since `not` will invert both "missing" and "false" to true.

Also, `has(container.securityContext.runAsUser) || has(container.securityContext.runAsNonRoot)` should be `has(container.securityContext.runAsUser) && has(container.securityContext.runAsNonRoot)` , since both conditions need to be satisfied.